test/e2e/framework/job/wait.go

```diff
@@ -82,14 +83,33 @@ func waitForJobPodsInPhase(ctx context.Context, c clientset.Interface, ns, jobNa
 // both conformance CI jobs with GA-only features and e2e CI jobs with all default-enabled features.
 // So, we need to skip "Complete" condition reason verifications in the e2e conformance test cases.
 func WaitForJobComplete(ctx context.Context, c clientset.Interface, ns, jobName string, reason *string, completions int32) error {
-	if err := wait.PollUntilContextTimeout(ctx, framework.Poll, JobTimeout, false, func(ctx context.Context) (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
-		if err != nil {
-			return false, err
+	// This function is called by HandleRetry, which will retry
+	// on transient API errors or stop polling in the case of other errors.
+	get := func(ctx context.Context) (*batchv1.Job, error) {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
+		if job != nil {
+			if isJobFailed(job) {
+				return nil, fmt.Errorf("job %q failed while waiting for its completion", klog.KObj(job))
+			}
+			return job, nil
+		}
+		return nil, err
```

Comment:
It is a better practice to check for an error first. Also, are we sure the job is nil when we get the error?

Reply:
yeah, you're right 🤦🏻 . Let me fix that

Reply:
thx!